curvefs/src/client/fuse_client.cpp

```diff
@@ -678,6 +812,246 @@ CURVEFS_ERROR FuseClient::FuseOpSetAttr(fuse_req_t req, fuse_ino_t ino,
+            &(inode->mutable_xattr()->find(XATTRFBYTES)->second),
+                std::to_string(fbytes), true);
+    } else {
+        return ret;
+    }
+
+    // all layer
+    uint64_t subdir = 0;
+    while (!iStack.empty()) {
+        ino = iStack.top();
+        iStack.pop();
+        std::list<Dentry> dentryList;
+        ret = dentryManager_->ListDentry(ino, &dentryList, limit, false);
+        if (CURVEFS_ERROR::OK != ret) {
+            LOG(ERROR) << "ListDentry failed, inodeId = " << ino
+                       << ", limit = " << limit << ", onlyDir = false";
+            return ret;
+        }
+        for (const auto &it : dentryList) {
+            if (it.type() == FsFileType::TYPE_DIRECTORY) {
+                iStack.emplace(it.inodeid());
+                subdir++;
+            }
+            inodeIds.emplace(it.inodeid());
+        }
+    }
+
+    VLOG(1) << "RcalculateSumInfo subdir_num = " << subdir
+            << ", entries_num = " << inodeIds.size();
+    attrs.clear();
+    ret = inodeManager_->BatchGetInodeAttr(inodeIds, &attrs);
+    if (ret == CURVEFS_ERROR::OK) {
+        uint64_t rfiles = 0;
+        uint64_t rsubdirs = 0;
+        uint64_t rentries = 0;
+        uint64_t rfbytes = 0;
+        for (const auto &it : attrs) {
+            if (it.type() == FsFileType::TYPE_DIRECTORY) {
+                rsubdirs++;
+            } else {
+                rfiles++;
+            }
+            rentries++;
+            rfbytes += it.length();
+        }
+        inode->mutable_xattr()->insert({XATTRRFILES, std::to_string(rfiles)});
+        inode->mutable_xattr()->insert({XATTRRSUBDIRS,
+            std::to_string(rsubdirs)});
+        inode->mutable_xattr()->insert({XATTRRENTRIES,
+            std::to_string(rentries)});
+        inode->mutable_xattr()->insert({XATTRRFBYTES,
+            std::to_string(rfbytes + inode->length())});
+    }
+
+    return ret;
+}
+
+CURVEFS_ERROR FuseClient::CalculateSumInfo(Inode *inode) {
+    std::stack<uint64_t> iStack;
+    std::set<uint64_t> inodeIds;
+    auto ino = inode->inodeid();
+    auto limit = option_.listDentryLimit;
+    iStack.emplace(ino);
+    while (!iStack.empty()) {
+        ino = iStack.top();
+        iStack.pop();
+        std::list<Dentry> dentryList;
+        auto ret = dentryManager_->ListDentry(ino, &dentryList, limit, true);
+        if (CURVEFS_ERROR::OK != ret) {
+            LOG(ERROR) << "ListDentry failed, inodeId = " << ino
+                       << ", limit = " << limit << ", onlyDir = true";
+            return ret;
+        }
+        for (const auto &it : dentryList) {
+            iStack.emplace(it.inodeid());
+            inodeIds.emplace(it.inodeid());
+        }
+    }
+
+    VLOG(1) << "CalculateSumInfo inode_num = " << inodeIds.size();
+    std::list<XAttr> xattrs;
+    auto ret = inodeManager_->BatchGetXAttr(inodeIds, &xattrs);
+    if (ret == CURVEFS_ERROR::OK) {
+        inode->mutable_xattr()->insert({XATTRRFILES,
```

Comment:
if key exist, insert can success?

Reply:
> if key exist, insert can success?

The curve.dir.r* have not recorded in dir xattr. Just pass the calculated value througth the temp inode xattr.